onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py

```diff
@@ -0,0 +1,474 @@
+    dynamic_axes = {}
+    for name in input_names + output_names:
+        if name in {"input_ids", "position_ids"}:
+            # shape is (batch_size, 1)
+            dynamic_axes[name] = {0: "batch_size"}
+        elif name == "attention_mask":
+            # shape is (batch_size, past_sequence_length + 1)
+            dynamic_axes[name] = {0: "batch_size", 1: "past_sequence_length + 1"}
+        elif "past" in name:
+            # shape is (batch_size, num_heads, past_sequence_length, head_size)
+            dynamic_axes[name] = {0: "batch_size", 2: "past_sequence_length"}
+        elif name == "logits":
+            # shape is (batch_size, 1, vocab_size)
+            dynamic_axes[name] = {0: "batch_size"}
+        elif "present" in name:
+            # shape is (batch_size, num_heads, past_sequence_length + 1, head_size)
+            dynamic_axes[name] = {0: "batch_size", 2: "past_sequence_length + 1"}
+        else:
+            raise Exception("Unknown input or output name found")
+    return dynamic_axes
+
+
+def save_onnx_model(onnx_model: onnx.ModelProto, output_path: str, data_path: str):
+    onnx.save(
+        onnx_model,
+        output_path,
+        save_as_external_data=True,
+        all_tensors_to_one_file=True,
+        location=data_path,
+        size_threshold=1024,
+        convert_attribute=False,
+    )
+
+
+# Notes:
+# 1) Dynamo export will not work automatically until this issue is resolved: https://github.com/microsoft/onnxscript/issues/493
+#
+# 2) Dynamo export will run manually if you set the ONNX file path to the same path that you use to save the model after export.
+# In other words, the value of `temp_path` should be set as the ONNX file path. You can open the issue in your browser to find
+# the location in ONNX Script where you have to make this change.
+#
+# Once the issue is resolved, we can modify the code below as follows for each export.
+#
+# Before:
+# temp_dir = args.output
+# temp_path = os.path.join(temp_dir, "temp.onnx")
+# ...
+# ...
+# ...
+# del onnx_model
+# os.system(f"rm {os.path.join(temp_dir, 'model.*')} && rm {os.path.join(temp_dir, '*.weight')} && rm {temp_path}")
+#
+#
+# After:
+# temp_dir = tempfile.TemporaryDirectory()
+# temp_path = os.path.join(temp_dir.name, "temp.onnx")
+# ...
+# ...
+# ...
+# del onnx_model
+# temp_dir.cleanup()
+#
+#
+# 3) Once exported, you will need to manually remove the external data that is saved to individual files. Do NOT delete the
+# file with the ".onnx.data" extension.
+def run_dynamo_export(args: argparse.Namespace, l_config: LlamaConfig, llama: LlamaForCausalLM):
+    from torch._dynamo import config
+
+    config.capture_scalar_outputs = True
+
+    # Export decoder_model.onnx
+    input_ids, attn_mask, pos_ids = get_model_inputs(l_config)
+    temp_dir = args.output  # tempfile.TemporaryDirectory()
+    temp_path = os.path.join(temp_dir, "temp.onnx")  # os.path.join(temp_dir.name, "temp.onnx")
```

Comment:
Maybe use [NamedTemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) would help. When used within a context manager, temp files are automatically deleted for you

You can refer to this [unit test](https://github.com/pytorch/pytorch/blob/5c3aae8385883eee03e42d3a409afcaf1b6e7959/test/onnx/test_fx_to_onnx_with_onnxruntime.py#L846) for an example

Reply:
The temp file would be deleted but I believe the external data files would not be removed. Additionally, until [support for saving large models in ONNX Script is resolved](https://github.com/microsoft/onnxscript/issues/493), the path to save the exported model (`temp_path`) currently has to match the specified path in the ONNX Script code. If a temp file is used, then these paths would not match before running the converter script.